flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/utils/ParquetSchemaConverter.java

```diff
@@ -143,6 +146,13 @@ private static Type convertToParquetType(
                         convertToParquetType(LIST_ELEMENT_NAME, arrayType.getElementType(), conf));
             case MAP:
                 MapType mapType = (MapType) type;
+                Preconditions.checkState(
+                        !mapType.getKeyType().isNullable(),
```

Comment:
I have concern here, does it will break the compatibility before?

Reply:
Hi @JingsongLi !

Yes, it does break backwards compatibility. After upgrade, jobs running with nullable map keys and multiset items in their schema will throw `InvalidStateException`. I do not see an alternative though as Parquet files these jobs produce are invalid and may not be readable by other tools.

Reply:
Can we automatically convert to not null? I understand that Flink SQL defaults to nullable.

Reply:
My reasoning about this issue is the following. 

Flink type system is fundamentally incompatible with Parquet type system. Flink allows nullable map keys, Parquet does not allow them.  Right now, Flink's `ParquetSchemaConverter` happily maps any Flink row type to Parquet message type, producing invalid Parquet files for incompatible schemas. Some external tools either silently return incorrect result or fail on these invalid Parquet files.

I see 2 possible solutions:
1. Do not allow to create tables with schema that is not compatible with Parquet spec (i.e., what's implemented in this patch).
2. Unconditionally specify that Parquet map key is `required` and throw in the writer if a record does have `null` key.

The main advantage of Solution 1 is that we correctly use Flink's type system guardrail against unsupported data types. This is what the type systems are for!
The main disadvantage of Solution 1 is that it is inconvenient to upgrade to because one'd have to update table schemas. While inconvenient, it does make the operator think whether that table can have invalid records.

The main advantage of Solution 2 is that only jobs producing invalid data will throw an exception, albeit at processing time. 
The disadvantages of this approach are:
1.  incompatible jobs will crashloop at processing time rather than at schema modeling time.
2. there will be additional per-record overhead for schemas with maps to ensure that they do not contain `null` keys. I can imagine this overhead being non-trivial, especially for complex schemas with multiple nesting levels.

For completeness, I did not bring up a solution where we unconditionally specify Parquet map key as `required`, but do not verify records in the writer, because it does not solve the problem at hand. With that appeoach, Flink still can produce invalid Parquet files, and external tooling will crash while reading them.

---

To sum up, I am fine with either of 2 solutions with the slight preference to Solution 1 because it does not introduce performance overhead and makes operators aware of the potential type incompatibility sooner than later.

What do you think?

Reply:
I tend to lean towards the second solution, because it would be a bit frustrating if the default type (Most users are too lazy to use not null) cannot be used.

Perhaps Parquet will report an error when encountering a null value? Is there no performance loss?

Reply:
It seems that when Spark SQL creates a Map type by default, the key is not null? Key cannot have null, this is its definition.

Reply:
Yes, Spark SQL Map does not allow null keys, that's why they do not have this issue. 

> Perhaps Parquet will report an error when encountering a null value?

Yes, Parquet writer will enter a crashloop on a record with null key.

Reply:
>  Is there no performance loss?

we'll have to identify maps in the schema and then check for each record that it does not have a null key. This may add up for complex nested schemas.

Reply:
I am happy to implement Solution 2 though.

Reply:
Sounds good to me.

Reply:
Done in https://github.com/apache/flink/pull/24956/commits/80b0bec76ee06bf8e48227d3a9f6c71ac0d3a8e6.